File: motor/coordinator/scheduler/runtime/workload_shm/layout.py
# Copyright (c) Huawei Technologies Co., Ltd. 2025-2026. All rights reserved.
# MindIE is licensed under Mulan PSL v2.
# You can use this software according to the terms and conditions of the Mulan PSL v2.
# You may obtain a copy of Mulan PSL v2 at:
#         http://license.coscl.org.cn/MulanPSL2
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
# EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
# MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
# See the Mulan PSL v2 for more details.

"""
Shared memory layout for workload data.
Header 64B + Entry 40B × N. Little-endian. SCHEMA_VERSION=5.
Control-plane membership uses header seqlock; per-slot active_tokens / isl /
cpu_hit_blocks are AtomicU64 CAS at offsets 16/24/32 (8-aligned). Sequence odd means writer
in progress (membership snapshot).
"""

import struct
from dataclasses import dataclass

# Magic: 4-byte format signature at start of shared memory header.
# "WKLD" (WorkLoad) as ASCII -> 0x57 0x4B 0x4C 0x44 = 0x574B4C44 (little-endian).
# Readers check this to ensure the buffer is our workload shm layout, not other data or corruption.
MAGIC = 0x574B4C44

# Schema version for layout compatibility (schema 5: per-slot CAS of tokens + overlay fields)
SCHEMA_VERSION = 5

# Role mapping: prefill=0, decode=1, hybrid=2, encode=3
ROLE_PREFILL = 0
ROLE_DECODE = 1
ROLE_HYBRID = 2
ROLE_ENCODE = 3

# Header: 64 bytes
# magic 4B, schema 2B, padding 2B, sequence 8B (seqlock), entry_count 4B, max_entries 4B,
# instance_version 8B (bumped when instance/endpoint set changes),
# heartbeat_sequence 8B (Scheduler bumps ~1/s),
# prefill_sequence 8B, decode_sequence 8B, hybrid_sequence 8B
HEADER_SIZE = 64
HEADER_FMT = "<I H H q I I Q Q Q Q Q"  # little-endian
HEARTBEAT_OFFSET = 32  # bytes 32-40: heartbeat_sequence (Q)
HEARTBEAT_STALE_SEC = 5.0  # If heartbeat unchanged for this long, Infer treats shm as stale

# Entry: 40 bytes (schema 5)
# instance_id 4B, endpoint_id 4B, role 1B, flags 1B, generation 2B, reserved 4B,
# active_tokens 8B at offset 16, isl 8B at 24, cpu_hit_blocks 8B at 32
# (all three 8-byte aligned for AtomicU64 CAS on aarch64).
ENTRY_SIZE = 40
ENTRY_FMT = "<i i B B H I d d d"

# Entry flag bits (must match workload_shm_rs/src/layout.rs).
FLAG_BLOCKED = 0b0000_0001
FLAG_VALID = 0b0000_0010

# Max number of (instance, endpoint) workload entries in shared memory. Not user-configurable.
DEFAULT_WORKLOAD_SHM_MAX_ENTRIES = 10240


@dataclass(frozen=True)
class WorkloadShmEntry:
    """Single workload entry (40 bytes). Used by pack_entry/unpack_entry and writer."""

    instance_id: int
    endpoint_id: int
    role: int
    active_tokens: float
    flags: int = 0
    generation: int = 0
    isl: float = 0.0
    cpu_hit_blocks: float = 0.0


@dataclass(frozen=True)
class WorkloadShmHeader:
    """Header fields for workload shared memory (64 bytes). Used by pack_header/unpack_header."""

    magic: int
    schema_version: int
    sequence: int
    entry_count: int
    max_entries: int
    instance_version: int = 0
    heartbeat_sequence: int = 0
    prefill_sequence: int = 0
    decode_sequence: int = 0
    hybrid_sequence: int = 0


def pack_header(header: WorkloadShmHeader) -> bytes:
    """Pack header into 64 bytes. Unsigned sequence fields are normalized to uint64."""
    instance_version = header.instance_version
    heartbeat_sequence = header.heartbeat_sequence
    prefill_sequence = header.prefill_sequence
    decode_sequence = header.decode_sequence
    hybrid_sequence = header.hybrid_sequence
    if instance_version < 0 or instance_version > (1 << 64) - 1:
        instance_version = instance_version % (1 << 64)
    if heartbeat_sequence < 0 or heartbeat_sequence > (1 << 64) - 1:
        heartbeat_sequence = heartbeat_sequence % (1 << 64)
    if prefill_sequence < 0 or prefill_sequence > (1 << 64) - 1:
        prefill_sequence = prefill_sequence % (1 << 64)
    if decode_sequence < 0 or decode_sequence > (1 << 64) - 1:
        decode_sequence = decode_sequence % (1 << 64)
    if hybrid_sequence < 0 or hybrid_sequence > (1 << 64) - 1:
        hybrid_sequence = hybrid_sequence % (1 << 64)
    return struct.pack(
        HEADER_FMT,
        header.magic,
        header.schema_version,
        0,  # padding
        header.sequence,
        header.entry_count,
        header.max_entries,
        instance_version,
        heartbeat_sequence,
        prefill_sequence,
        decode_sequence,
        hybrid_sequence,
    )


def unpack_header(buf: memoryview) -> WorkloadShmHeader:
    """Parse 64-byte header from buffer. Returns WorkloadShmHeader."""
    if len(buf) < HEADER_SIZE:
        raise ValueError(f"Buffer too small for header: {len(buf)} < {HEADER_SIZE}")
    t = struct.unpack(HEADER_FMT, buf[:HEADER_SIZE])
    return WorkloadShmHeader(
        magic=t[0],
        schema_version=t[1],
        sequence=t[3],
        entry_count=t[4],
        max_entries=t[5],
        instance_version=t[6],
        heartbeat_sequence=t[7],
        prefill_sequence=t[8],
        decode_sequence=t[9],
        hybrid_sequence=t[10],
    )


def pack_entry(entry: WorkloadShmEntry) -> bytes:
    """Pack single entry into 40 bytes."""
    return struct.pack(
        ENTRY_FMT,
        entry.instance_id,
        entry.endpoint_id,
        entry.role,
        entry.flags,
        entry.generation,
        0,  # reserved
        entry.active_tokens,
        entry.isl,
        entry.cpu_hit_blocks,
    )


def unpack_entry(buf: memoryview, slot: int) -> WorkloadShmEntry:
    """Unpack entry at slot. Returns WorkloadShmEntry."""
    offset = HEADER_SIZE + slot * ENTRY_SIZE
    if offset + ENTRY_SIZE > len(buf):
        raise ValueError(f"Entry slot {slot} out of range")
    t = struct.unpack(ENTRY_FMT, buf[offset : offset + ENTRY_SIZE])
    return WorkloadShmEntry(
        instance_id=t[0],
        endpoint_id=t[1],
        role=t[2],
        flags=t[3],
        generation=t[4],
        active_tokens=t[6],
        isl=t[7],
        cpu_hit_blocks=t[8],
    )


def total_size(max_entries: int) -> int:
    """Total shared memory size in bytes."""
    return HEADER_SIZE + max_entries * ENTRY_SIZE

File: motor/coordinator/workload_shm_rs/src/layout.rs
// Copyright (c) Huawei Technologies Co., Ltd. 2025-2026. All rights reserved.
// MindIE is licensed under Mulan PSL v2.
// You can use this software according to the terms and conditions of the Mulan PSL v2.
// You may obtain a copy of Mulan PSL v2 at:
//         http://license.coscl.org.cn/MulanPSL2
// THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
// EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
// MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
// See the Mulan PSL v2 for more details.

//! Byte layout for the workload shared-memory segment.
//!
//! This mirrors, byte-for-byte, the Python layout in
//! `motor/coordinator/scheduler/runtime/workload_shm/layout.py` (SCHEMA_VERSION 5):
//! a 64-byte header followed by N 40-byte entries, little-endian. Membership is seqlock-
//! published by Mgmt; per-slot `active_tokens` / `isl` / `cpu_hit_blocks` are
//! AtomicU64 CAS'd by Infer Workers.

/// Magic "WKLD" (0x57 0x4B 0x4C 0x44) little-endian.
pub const MAGIC: u32 = 0x574B_4C44;
/// Layout schema version. Must match the Python reader/owner.
pub const SCHEMA_VERSION: u16 = 5;

pub const HEADER_SIZE: usize = 64;
pub const ENTRY_SIZE: usize = 40;
pub const DEFAULT_MAX_ENTRIES: u32 = 10240;

// Header field byte offsets (see layout.py HEADER_FMT "<I H H q I I Q Q Q Q Q").
pub const OFF_MAGIC: usize = 0; // u32
pub const OFF_SCHEMA: usize = 4; // u16
pub const OFF_SEQUENCE: usize = 8; // i64 (seqlock; odd = write in progress)
pub const OFF_ENTRY_COUNT: usize = 16; // u32
pub const OFF_MAX_ENTRIES: usize = 20; // u32
pub const OFF_INSTANCE_VERSION: usize = 24; // u64
pub const OFF_HEARTBEAT: usize = 32; // u64
pub const OFF_PREFILL_SEQ: usize = 40; // u64
pub const OFF_DECODE_SEQ: usize = 48; // u64
pub const OFF_HYBRID_SEQ: usize = 56; // u64

// shm role bytes (layout.py: prefill=0, decode=1, hybrid=2, encode=3).
pub const ROLE_PREFILL: u8 = 0;
pub const ROLE_DECODE: u8 = 1;
pub const ROLE_HYBRID: u8 = 2;
pub const ROLE_ENCODE: u8 = 3;

// ---------------------------------------------------------------------------
// Schema 5: per-slot atomic CAS layout. Header is 64B; seqlock covers only membership
// (token/overlay CAS does NOT bump it), so readers must atomic-load ledger fields every pass.
// ---------------------------------------------------------------------------

// Entry field byte offsets within a 40-byte slot.
//
// The three f64 ledgers sit at 16/24/32 so that, with an 8-aligned segment base and a 40B
// stride, every slot's atomics are 8-byte aligned (mandatory on aarch64 / Ascend hosts).
pub const ENTRY_V4_OFF_INSTANCE_ID: usize = 0; // i32 (written on snapshot only)
pub const ENTRY_V4_OFF_ENDPOINT_ID: usize = 4; // i32 (written on snapshot only)
pub const ENTRY_V4_OFF_ROLE: usize = 8; // u8 (written on snapshot only)
pub const ENTRY_V4_OFF_FLAGS: usize = 9; // u8, AtomicU8 (BLOCKED / VALID)
pub const ENTRY_V4_OFF_GENERATION: usize = 10; // u16 (written on snapshot only; ABA guard)
pub const ENTRY_V4_OFF_RESERVED: usize = 12; // u32
pub const ENTRY_V4_OFF_ACTIVE_TOKENS: usize = 16; // u64 (f64::to_bits), AtomicU64
pub const ENTRY_V5_OFF_ISL: usize = 24; // u64 (f64::to_bits), AtomicU64
pub const ENTRY_V5_OFF_CPU_HIT_BLOCKS: usize = 32; // u64 (f64::to_bits), AtomicU64

// Entry flags bits.
pub const FLAG_BLOCKED: u8 = 0b0000_0001; // circuit-breaker OPEN: allocate CAS must refuse
pub const FLAG_VALID: u8 = 0b0000_0010; // slot holds a live (instance, endpoint)

const _: () = assert!(ENTRY_V5_OFF_CPU_HIT_BLOCKS + 8 <= ENTRY_SIZE);

/// Total segment size in bytes for `max_entries` slots.
pub fn total_size(max_entries: u32) -> usize {
    HEADER_SIZE + (max_entries as usize) * ENTRY_SIZE
}

/// Byte offset of the given slot's entry.
pub fn entry_offset(slot: u32) -> usize {
    HEADER_SIZE + (slot as usize) * ENTRY_SIZE
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn sizes_match_python_layout() {
        assert_eq!(HEADER_SIZE, 64);
        assert_eq!(ENTRY_SIZE, 40);
        assert_eq!(total_size(10240), 64 + 10240 * 40);
        assert_eq!(entry_offset(0), 64);
        assert_eq!(entry_offset(1), 104);
    }

    #[test]
    fn schema5_ledgers_are_8_byte_aligned_for_every_slot() {
        for slot in 0..1024u32 {
            for field in [
                ENTRY_V4_OFF_ACTIVE_TOKENS,
                ENTRY_V5_OFF_ISL,
                ENTRY_V5_OFF_CPU_HIT_BLOCKS,
            ] {
                let off = entry_offset(slot) + field;
                assert_eq!(off % 8, 0, "slot {slot} field {field} offset {off} not 8-aligned");
            }
        }
    }
}
